internal/persistence/postgres/pool.go

```diff
@@ -0,0 +1,154 @@
+	MinConns int32
+
+	// MaxConnLifetime is the duration since creation after which a connection will
+	// be automatically closed. Default to `persistence.DefaultMaxConnLifetime`.
+	MaxConnLifetime time.Duration
+
+	// MaxConnIdleTime is the duration after which an idle connection will be automatically
+	// closed by the health check. Default to `persistence.DefaultMaxConnIdleTime`.
+	MaxConnIdleTime time.Duration
+
+	// HealthCheckPeriod is the duration between checks of the health of
+	// idle connections. Default to `persistence.DefaultHealthCheckPeriod`.
+	HealthCheckPeriod time.Duration
+
+	// ReadOnly indicates the pool will be used for read only operations.
+	ReadOnly bool
+}
+
+// Validate ensures the provided Postgres.Pool options are a valid configuration.
+func (opts *PoolOpts) Validate() error {
+	// by validating these values, we avoid panics from pgxpool at connection
+	if opts.MaxConns < 1 {
+		return fmt.Errorf("invalid Pool.MaxConns value: '%d', should be greater than 0", opts.MaxConns)
+	}
+
+	if opts.HealthCheckPeriod < 1 {
+		return fmt.Errorf("invalid Pool.HealthCheckPeriod value: '%d', should be a positive duration", opts.HealthCheckPeriod)
+	}
+
+	return nil
+}
+
+func newPostgresPool(opts Opts, logger *zap.Logger) (Pool, error) {
+	var dbPool Pool
+	var err error
+
+	if poolOpenFunc, present := pools[opts.Pool.Name]; present {
+		dbPool, err = poolOpenFunc(opts, logger)
+		if err != nil {
+			return nil, err
+		}
+		return dbPool, nil
+	}
+
+	return nil, fmt.Errorf("invalid postgres pool '%s'", opts.Pool.Name)
+}
+
+// NewPgxPoolConfig creates and configure a new pgxpool.Config with the given options.
+func NewPgxPoolConfig(opts Opts, logger *zap.Logger) (*pgxpool.Config, error) {
+	dsn, err := opts.DSN(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := pgxpool.ParseConfig(dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	config.MaxConns = opts.Pool.MaxConns
+	config.MinConns = opts.Pool.MinConns
+	config.MaxConnLifetime = opts.Pool.MaxConnLifetime
+	config.MaxConnIdleTime = opts.Pool.MaxConnIdleTime
+	config.HealthCheckPeriod = opts.Pool.HealthCheckPeriod
+
+	return config, nil
+}
+
+func newPgxPool(opts Opts, logger *zap.Logger) (Pool, error) {
+	logger.Info("using Postgres PGX Pool")
+	ctx := context.Background()
+
+	pgxPoolConfig, err := NewPgxPoolConfig(opts, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	dbPool, err := pgxpool.ConnectConfig(ctx, pgxPoolConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	err = dbPool.Ping(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return dbPool, nil
+}
+
+// RegisterPool adds a `Pool` implementation to the postgres configuration.
+func RegisterPool(poolName string, poolOpenFunc PoolOpenFunc) error {
+	if _, present := pools[poolName]; present {
+		return fmt.Errorf("pool name '%s' already exists", poolName)
+	}
+	pools[poolName] = poolOpenFunc
+
+	return nil
+}
+
+func init() {
+	pools[DefaultPool] = newPgxPool
```

Comment:
I don't understand why we're doing this. Can we throw some documentation on why we need to store all the registered pools? I imagine it's for the healthcheck stuff to ensure we can easily look up all pools at once, but perhaps I'm wrong.

Reply:
not what what you mean, it's only to be able to select the proper pool implementation by name from the configuration. e.g:
`  postgres:
    hostname: localhost
    port: 5432
    db_name: koko
    user: koko
    password: koko
    pool:
      name: aurora
      health_check_period: 10s
`

so a `pool_name:pool_init_func` registration.

Reply:
Err, yeah you can forget this, I see what it's doing now. I should have deleted this comment.